File: Backend/backend/src/main/java/com/example/backend/services/dbservices/HabitsService.java
package com.example.backend.services.dbservices;

import com.example.backend.DTOs.HabitsDTO;
import com.example.backend.DTOs.TrackedDaysDTO;
import com.example.backend.components.habitscomponents.HabitsComponentCRUD;
import com.example.backend.components.habitscomponents.HabitsMarkingComponent;
import com.example.backend.components.habitscomponents.HabitsTrackingComponent;
import com.example.backend.components.interfaces.ComponentCrud;
import com.example.backend.components.interfaces.ComponentEntityCRUD;
import com.example.backend.components.interfaces.MarkingInterface;
import com.example.backend.components.usercomponents.UserAuthComponent;
import com.example.backend.entities.Habits;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class HabitsService implements DbaServiceInterface<HabitsDTO> {
    private final ComponentCrud<HabitsDTO> habitsComponentCRUD;
    private final ComponentEntityCRUD<Habits> habitsComponentEntityCRUD;
    private final MarkingInterface habitsMarkingComponent;
    private final UserAuthComponent userAuthComponent;
    private final HabitsTrackingComponent habitsTrackingComponent;

    @Override
    public List<HabitsDTO> get(Authentication auth) {
        List<HabitsDTO> list = habitsComponentCRUD.get(userAuthComponent.getUserByAuthorities(auth));
        list.forEach(habitsDTO -> habitsDTO.setTracks(getAllTracks(habitsDTO.getId())));
        return list;
    }

    @Override
    public HabitsDTO getById(Long id) {
        return habitsComponentCRUD.getById(id);
    }

    @Override
    public HabitsDTO create(HabitsDTO habitsDTO, Authentication auth) {
        return habitsComponentCRUD.create(habitsDTO, userAuthComponent.getUserByAuthorities(auth));
    }

    @Override
    public HabitsDTO update(HabitsDTO habitsDTO, Long id, Authentication auth) {
        return habitsComponentCRUD.update(habitsDTO, id, userAuthComponent.getUserByAuthorities(auth));
    }

    @Override
    public void delete(Long id) {
        habitsComponentCRUD.delete(id);
    }

    public void mark(Long id) {
        habitsMarkingComponent.markIt(id);
    }

    public void unmark(Long id) {
        habitsMarkingComponent.unmarkIt(id);
    }

    private List<TrackedDaysDTO> getAllTracks(Long id) {

        return habitsTrackingComponent.getAllTracksByHabit(habitsComponentEntityCRUD.getEntityById(id));
    }
}

File: Frontend/front/src/services/HabitsService.js
import axios from "axios";

const HABITS_API_URL = "http://localhost:8080/api/habits"

class HabitsService {
    getAll() {
        return axios.get(HABITS_API_URL, {withCredentials: true})
    }

    create(HabitDTO) {
        return axios.post(HABITS_API_URL, HabitDTO, {withCredentials: true})
    }

    update(HabitDTO, id) {
        return axios.put(HABITS_API_URL + '/' + id, HabitDTO, {withCredentials: true})
    }

    delete(id) {
        return axios.delete(HABITS_API_URL + '/' + id, {withCredentials: true})
    }

    mark(id) {
        let body = true
        return axios.patch(HABITS_API_URL + '/mark/' + id, body, {withCredentials: true})
    }

    unmark(id) {
        let body = false
        return axios.patch(HABITS_API_URL + '/unmark/' + id, body, {withCredentials: true})
    }
}

export default new HabitsService()